### DISCUSSION TITLE
Code to parse L2 txs from L1 calldata?

### DISCUSSION THREAD

**Discussion by bkellerman** [ethereum-optimism/developers#98] [General]
### Discussion Topic

I am following the spec [here](https://github.com/ethereum-optimism/optimism/blob/develop/specs/derivation.md#frame-format) but am only able to parse the channel frame, not the frame data specifically.

Is there code or an sdk somewhere that can extract full L2 txs from L1 calldata?

Thanks.

**Comment by sbvegan** [comment]
I think you can use the `op-node` [batch_decoder](https://github.com/ethereum-optimism/optimism/tree/develop/op-node/cmd/batch_decoder) tool as a reference. Let me know what you think.

**Reply by sbvegan** [reply]
Hey @bkellerman, just following up to see if this tool helps or if you're looking for something different.

**Reply by bkellerman** [reply]
The batch_decoder is helpful to demonstrate reassembly of channels but doesn't decode the L2 txs.

**Reply by sbvegan** [reply]
The code to accomplish it on the L1 doesn't exist at the moment. You can use the [code in the derivation pipeline](- https://github.com/ethereum-optimism/optimism/blob/afbe2e8981c0f818ae8d06d30d749ede91974dcd/op-node/rollup/derive/attributes_queue.go#L78) because the batch contains all of the transactions. It does require contextual knowledge to map the transactions. 

Can you tell us more about your use case? We might be able to provide better support with additional context.
